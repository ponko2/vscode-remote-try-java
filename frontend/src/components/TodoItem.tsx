import { cva } from "class-variance-authority";
import clsx from "clsx";
import type React from "react";
import { useEffect, useRef, useState } from "react";
import type { components } from "../api/schema";
import { TodoButton } from "./TodoButton";

type Props = {
  todo: components["schemas"]["TodoResponse"];
  deleteTodo: (id: string) => void;
  toggleTodo: (data: components["schemas"]["TodoResponse"]) => void;
  updateTodo: (data: components["schemas"]["TodoResponse"]) => void;
};

export function TodoItem({ todo, deleteTodo, toggleTodo, updateTodo }: Props) {
  const [editing, setEditing] = useState(false);
  const [value, setValue] = useState(todo.title);
  const ref = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (editing) {
      ref.current?.focus();
    }
  }, [editing]);

  function saveTodo(data: components["schemas"]["TodoResponse"]) {
    if (data.title.length === 0) {
      deleteTodo(data.id);
    } else {
      updateTodo(data);
    }
    setEditing(false);
  }

  function handleBlur(event: React.FocusEvent<HTMLInputElement>) {
    saveTodo({ ...todo, title: event.target.value });
  }

  function handleChange(event: React.ChangeEvent<HTMLInputElement>) {
    setValue(event.target.value);
  }

  function handleKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
    const title = event.currentTarget.value.trim();
    if (event.key === "Enter") {
      saveTodo({ ...todo, title });
    }
  }

  function handleDoubleClick() {
    setEditing(true);
  }

  const list = cva(["relative", "text-2xl", "h-16"], {
    variants: {
      intent: {
        primary: ["group"],
        editing: ["pl-11"],
      },
    },
    defaultVariants: {
      intent: "primary",
    },
  });

  if (editing) {
    return (
      <li className={list({ intent: "editing" })}>
        <input
          className={clsx(
            "size-full",
            "border",
            "border-neutral-400",
            "px-4",
            "py-3",
            "shadow-inner",
            "focus:shadow",
            "focus:shadow-red-400",
            "focus:outline-none",
          )}
          onBlur={handleBlur}
          onChange={handleChange}
          onKeyDown={handleKeyDown}
          ref={ref}
          type="text"
          value={value}
        />
      </li>
    );
  }
  return (
    <li className={list()}>
      <input
        checked={todo.completed}
        className={clsx(
          "peer",
          "absolute",
          "inset-y-0",
          "my-auto",
          "size-12",
          "appearance-none",
          "outline-none",
        )}
        onChange={() => toggleTodo(todo)}
        type="checkbox"
      />
      <label
        className={clsx(
          "block",
          "h-full",
          "break-words",
          "bg-unchecked",
          "bg-left",
          "bg-no-repeat",
          "py-4",
          "pl-14",
          "pr-4",
          "font-normal",
          "leading-tight",
          "text-neutral-700",
          "transition-colors",
          "duration-500",
          "peer-checked:bg-checked",
          "peer-checked:text-neutral-400",
          "peer-checked:line-through",
          "peer-focus:shadow",
          "peer-focus:shadow-red-400",
          "peer-focus:outline-none",
        )}
        onDoubleClick={handleDoubleClick}
      >
        {todo.title}
      </label>
      <TodoButton
        className={clsx(
          "absolute",
          "inset-y-0",
          "right-2.5",
          "my-auto",
          "hidden",
          "size-10",
          "text-3xl",
          "text-neutral-400",
          "transition-colors",
          "duration-200",
          "ease-out",
          "after:block",
          "after:h-full",
          "after:content-['Ã—']",
          "hover:text-red-400",
          "focus:text-red-400",
          "group-hover:block",
        )}
        onClick={() => deleteTodo(todo.id)}
      />
    </li>
  );
}
