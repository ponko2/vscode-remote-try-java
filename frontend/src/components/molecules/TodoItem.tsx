import React, { useState } from "react";
import { TodoButton } from "../../components/atoms/TodoButton";
import { TodoPatchRequestModel, TodoResponseModel } from "../../models/todos";
import styles from "./TodoItem.module.css";

type Props = {
  todo: TodoResponseModel;
  editTodo: (data: TodoPatchRequestModel) => void;
  deleteTodo: (id: string) => void;
  toggleTodo: (data: TodoPatchRequestModel) => void;
};

export const TodoItem = ({
  todo,
  editTodo,
  deleteTodo,
  toggleTodo,
}: Props): JSX.Element => {
  const [editing, setEditing] = useState(false);
  const [value, setValue] = useState(todo.title);

  function saveTodo(data: TodoPatchRequestModel) {
    if (data.title.length === 0) {
      deleteTodo(data.id);
    } else {
      editTodo(data);
    }
    setEditing(false);
  }

  function handleBlur(event: React.FocusEvent<HTMLInputElement>) {
    saveTodo({ ...todo, title: event.target.value });
  }

  function handleChange(event: React.ChangeEvent<HTMLInputElement>) {
    setValue(event.target.value);
  }

  function handleKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
    const newValue = event.currentTarget.value.trim();
    if (event.key === "Enter") {
      saveTodo({ ...todo, title: newValue });
    }
  }

  function handleDoubleClick() {
    setEditing(true);
  }

  const element = (() => {
    if (editing) {
      return (
        <input
          type="text"
          value={value}
          className={styles.edit}
          onBlur={handleBlur}
          onChange={handleChange}
          onKeyDown={handleKeyDown}
        />
      );
    }
    return (
      <div className={styles.view}>
        <input
          className={styles.toggle}
          type="checkbox"
          checked={todo.completed}
          onChange={() => toggleTodo(todo)}
        />
        <label onDoubleClick={handleDoubleClick}>{todo.title}</label>
        <TodoButton
          className={styles.destroy}
          onClick={() => deleteTodo(todo.id)}
        />
      </div>
    );
  })();

  return <li className={styles.todoItem}>{element}</li>;
};
