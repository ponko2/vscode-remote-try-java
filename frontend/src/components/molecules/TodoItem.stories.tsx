import { action } from "@storybook/addon-actions";
import type { Meta, StoryObj } from "@storybook/react";
import { TodoItem } from "./TodoItem";

const meta = {
  component: TodoItem,
  decorators: [
    (Story) => (
      <ul style={{ margin: 0, padding: 0, listStyle: "none" }}>
        <Story />
      </ul>
    ),
  ],
} satisfies Meta<typeof TodoItem>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Basic = {
  args: {
    editTodo: action("editTodo"),
    deleteTodo: action("deleteTodo"),
    toggleTodo: action("toggleTodo"),
    todo: {
      id: "01FYH5XVSNVSXTSGB8KB858REF",
      title: "Hello, World!!",
      completed: false,
    },
  },
} satisfies Story;

export const Completed = {
  args: {
    ...Basic.args,
    todo: {
      id: "01FYH5XVSNVSXTSGB8KB858REF",
      title: "Hello, World!!",
      completed: true,
    },
  },
} satisfies Story;
