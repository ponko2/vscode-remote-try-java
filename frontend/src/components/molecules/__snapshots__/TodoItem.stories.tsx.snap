// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/molecules/TodoItem Basic smoke-test 1`] = `
<ul style="margin: 0px; padding: 0px; list-style: none;">
  <li class="_todoItem_17lsu_1">
    <div class="_view_17lsu_69">
      <input class="_toggle_17lsu_34"
             type="checkbox"
      >
      <label>
        Hello, World!!
      </label>
      <button class="_button_z78gq_1 _destroy_17lsu_79">
      </button>
    </div>
  </li>
</ul>
`;

exports[`components/molecules/TodoItem Completed smoke-test 1`] = `
<ul style="margin: 0px; padding: 0px; list-style: none;">
  <li class="_todoItem_17lsu_1">
    <div class="_view_17lsu_69">
      <input class="_toggle_17lsu_34"
             type="checkbox"
             checked
      >
      <label>
        Hello, World!!
      </label>
      <button class="_button_z78gq_1 _destroy_17lsu_79">
      </button>
    </div>
  </li>
</ul>
`;
